pkg/informer/informer.go

```diff
@@ -126,108 +175,126 @@ func NewDynamicDiscoverySharedInformerFactory(
+// those types.
+// It offers an additional `Cluster()` method that returns a new shared informer factory
+// based on the main informer factory, but scoped for a given logical cluster.
+type DiscoveringDynamicSharedInformerFactory struct {
+	*GenericDiscoveringDynamicSharedInformerFactory[kcpcache.ScopeableSharedIndexInformer, kcpcache.GenericClusterLister, kcpinformers.GenericClusterInformer]
 }
 
-func (d *DynamicDiscoverySharedInformerFactory) Cluster(cluster logicalcluster.Name) kcpkubernetesinformers.ScopedDynamicSharedInformerFactory {
-	return &scopedDynamicDiscoverySharedInformerFactory{
-		DynamicDiscoverySharedInformerFactory: d,
-		cluster:                               cluster,
+func (d *DiscoveringDynamicSharedInformerFactory) Cluster(cluster logicalcluster.Name) kcpinformers.ScopedDynamicSharedInformerFactory {
+	return &scopedDiscoveringDynamicSharedInformerFactory{
+		DiscoveringDynamicSharedInformerFactory: d,
+		cluster:                                 cluster,
 	}
 }
 
-type scopedDynamicDiscoverySharedInformerFactory struct {
-	*DynamicDiscoverySharedInformerFactory
+type scopedDiscoveringDynamicSharedInformerFactory struct {
+	*DiscoveringDynamicSharedInformerFactory
 	cluster logicalcluster.Name
 }
 
 // ForResource returns the GenericInformer for gvr, creating it if needed. The GenericInformer must be started
-// by calling Start on the DynamicDiscoverySharedInformerFactory before the GenericInformer can be used.
-func (d *scopedDynamicDiscoverySharedInformerFactory) ForResource(gvr schema.GroupVersionResource) (informers.GenericInformer, error) {
-	clusterInformer, err := d.DynamicDiscoverySharedInformerFactory.ForResource(gvr)
+// by calling Start on the DiscoveringDynamicSharedInformerFactory before the GenericInformer can be used.
+func (d *scopedDiscoveringDynamicSharedInformerFactory) ForResource(gvr schema.GroupVersionResource) (informers.GenericInformer, error) {
+	clusterInformer, err := d.DiscoveringDynamicSharedInformerFactory.ForResource(gvr)
 	if err != nil {
 		return nil, err
 	}
 	return clusterInformer.Cluster(d.cluster), nil
 }
 
 // ForResource returns the GenericInformer for gvr, creating it if needed. The GenericInformer must be started
-// by calling Start on the DynamicDiscoverySharedInformerFactory before the GenericInformer can be used.
-func (d *DynamicDiscoverySharedInformerFactory) ForResource(gvr schema.GroupVersionResource) (kcpkubernetesinformers.GenericClusterInformer, error) {
+// by calling Start on the GenericDiscoveringDynamicSharedInformerFactory before the GenericInformer can be used.
+func (d *GenericDiscoveringDynamicSharedInformerFactory[Informer, Lister, GenericInformer]) ForResource(gvr schema.GroupVersionResource) (GenericInformer, error) {
 	// See if we already have it
 	d.informersLock.RLock()
 	inf := d.informers[gvr]
 	d.informersLock.RUnlock()
 
-	if inf != nil {
+	if (genericInformerBase[Informer, Lister])(inf) != nil {
```

Comment:
surprising type cast

Reply:
This is required because `nil` is untyped.

Doing:

```golang
if inf != nil {
```

Raises the following error: `invalid operation: cannot compare inf != nil (mismatched types GenericInformer and untyped nil)`.